vm/src/interpreter.cpp

```diff
@@ -106,51 +118,71 @@ void run(constant_pool pool, int ip, const char *bytes, size_t size) {
             char local_index = *(bytes + ip + 1);
             ip += 2;
             // Push the local onto the stack
-            *(int64_t *)(stack + stack_frame) = *(int64_t *)(locals + local_frame + local_index * 8);
-            stack_frame += 8;
+            stack.push_int(*(int64_t *)(locals + local_frame + local_index * 8));
             break;
         }
         case OP_SET_LOCAL: {
             // Read the 1 byte local index
             char index = *(bytes + ip + 1);
             ip += 2;
             // Pop the value from the stack
-            stack_frame -= 8;
+            int64_t value = stack.pop_int();
             // Set the local
-            *(int64_t *)(locals + local_frame + index * 8) = *(int64_t *)(stack + stack_frame);
-            break;
-        }
-        case OP_POP: {
-            // Pop the value from the stack
-            stack_frame -= 8;
-            ip += 1;
+            *(int64_t *)(locals + local_frame + index * 8) = value;
             break;
         }
         case OP_INT_TO_STR: {
-            // convert last local from stack to str
-            // the local in stack
-            auto* stack_local = (int64_t *)(stack + stack_frame - 8);
+            int64_t value = stack.pop_int();
 
-            int64_t value = *stack_local;
-            append_str_value(value, stack_local, pool);
+            int64_t string_ref = append_str_value(value, pool);
+            stack.push_string_constant_ref(string_ref);
 
             // goto next operation
             ip++;
             break;
         }
         case OP_FLOAT_TO_STR: {
-            // convert last local from stack to str
-            // the local in stack
-            auto* stack_local = (int64_t *)(stack + stack_frame - 8);
+            double value = stack.pop_double();
+
+            int64_t string_ref = append_str_value(value, pool);
+            stack.push_string_constant_ref(string_ref);
 
-            auto value = *(double *)(stack + stack_frame - 8);
-            append_str_value(value, stack_local, pool);
             // goto next operation
             ip++;
             break;
         }
+        case OP_IF_NOT_JUMP:
+        case OP_IF_JUMP: {
+            int64_t value = stack.pop_int();
+            size_t then_branch = ntohl(*(size_t *)(bytes + ip + sizeof(char)));
+            // as we are in a shell interpreter thus
```

Comment:
Are we considering that an exit code a boolean as it is? The boolean type also exists, and there must be only one representation.